hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java

```diff
@@ -79,7 +86,10 @@ public void addNode(DatanodeDetails datanodeDetails, NodeStatus nodeStatus)
       if (nodeMap.containsKey(id)) {
         throw new NodeAlreadyExistsException("Node UUID: " + id);
       }
-      nodeMap.put(id, new DatanodeInfo(datanodeDetails, nodeStatus));
+      DatanodeInfo datanodeInfo = new DatanodeInfo(datanodeDetails, nodeStatus);
+      // Make sure nodes in topology tree has all DN info.
+      clusterMap.add(datanodeInfo);
+      nodeMap.put(id, datanodeInfo);
```

Comment:
The node is already added to the cluster map on registration in `SCMNodeManager.register(...)`:

```
  @Override
  public RegisteredCommand register(
      DatanodeDetails datanodeDetails, NodeReportProto nodeReport,
      PipelineReportsProto pipelineReportsProto) {
      ...
        networkLocation = nodeResolve(dnsName);
        if (networkLocation != null) {
          datanodeDetails.setNetworkLocation(networkLocation);
        }

        clusterMap.add(datanodeDetails);
        nodeStateManager.addNode(datanodeDetails);
    ...
```

Why do we need to add it again here?

Reply:
The add to topology action during SCMNodeManager.register(...) is removed in this patch.  

The goal is to make sure the node object added into topology is of DatanodeInfo type, so that when a node is chosen through topology API, this node will have the operation state information already, to save an extra call of NodeManager to get this information.

Reply:
@nandakumar131 @bshashikant What do you guys think about this change (putting the DatanodeInfo into the topology map)? I remember we talked about something like this with decommission, and concluded we were happier if the DatanodeDetails reference was only stored in a single place (nodeManager) and anything that needs it should request it from there. Our fear was that some future change could replace the DatanodeInfo instance stored in one place and not in the other.

Reply:
SCMContainerPlacementRackAware already has a NodeManager instance passed into it. I think this change would be cleaner and safer if we just used NodeManager to lookup the NodeStatus. Then all the locks are obeyed and DatanodeInfo is stored only in a single place.

The problem then is that the NetworkToplogy instance holds only a Node instance. But we need a DatanodeDetails instance (which extends Node) to query NodeManager to get the NodeStatus.